ntp-proto/src/record.rs

```diff
@@ -0,0 +1,320 @@
+            }
+            6 => {
+                let str_data = read_bytes_exact(reader, record_len)?;
+                match String::from_utf8(str_data) {
+                    Ok(name) => Record::Server { critical, name },
+                    Err(e) => Record::Unknown {
+                        record_type,
+                        critical,
+                        data: e.into_bytes(),
+                    },
+                }
+            }
+            7 if record_len == 2 => Record::Port {
+                critical,
+                port: read_u16_be(reader)?,
+            },
+            _ => Record::Unknown {
+                record_type,
+                critical,
+                data: read_bytes_exact(reader, record_len)?,
+            },
+        })
+    }
+
+    pub fn write<A: Write>(&self, writer: &mut A) -> std::io::Result<()> {
+        // error out early when the record is invalid
+        self.validate()?;
+
+        // all messages start with the record type
+        let record_type = self.record_type() | ((self.is_critical() as u16) << 15);
+        writer.write_all(&record_type.to_be_bytes())?;
+
+        let size_of_u16 = std::mem::size_of::<u16>() as u16;
+        match self {
+            Record::EndOfMessage => {
+                writer.write_all(&0_u16.to_be_bytes())?;
+            }
+            Record::Unknown { data, .. } => {
+                writer.write_all(&(data.len() as u16).to_be_bytes())?;
+                writer.write_all(data)?;
+            }
+            Record::NextProtocol { protocol_ids } => {
+                let length = size_of_u16 * protocol_ids.len() as u16;
+                writer.write_all(&length.to_be_bytes())?;
+
+                for id in protocol_ids {
+                    writer.write_all(&id.to_be_bytes())?;
+                }
+            }
+            Record::Error { errorcode } => {
+                writer.write_all(&size_of_u16.to_be_bytes())?;
+                writer.write_all(&errorcode.to_be_bytes())?;
+            }
+            Record::Warning { warningcode } => {
+                writer.write_all(&size_of_u16.to_be_bytes())?;
+                writer.write_all(&warningcode.to_be_bytes())?;
+            }
+            Record::AeadAlgorithm { algorithm_ids, .. } => {
+                let length = size_of_u16 * algorithm_ids.len() as u16;
+                writer.write_all(&length.to_be_bytes())?;
+
+                for id in algorithm_ids {
+                    writer.write_all(&id.to_be_bytes())?;
+                }
+            }
+            Record::NewCookie { cookie_data } => {
+                let length = cookie_data.len() as u16;
+                writer.write_all(&length.to_be_bytes())?;
+
+                writer.write_all(cookie_data)?;
+            }
+            Record::Server { name, .. } => {
+                let length = name.len() as u16;
+                writer.write_all(&length.to_be_bytes())?;
+
+                writer.write_all(name.as_bytes())?;
```

Comment:
Again, spec says name should only consist of ascii, should we enforce that here?